homeassistant/components/unifi/entity.py

```diff
@@ -0,0 +1,176 @@
+"""UniFi entity representation."""
+from __future__ import annotations
+
+from collections.abc import Callable
+from dataclasses import dataclass
+from typing import Generic, TypeVar
+
+import aiounifi
+from aiounifi.interfaces.api_handlers import CallbackType, ItemEvent, UnsubscribeType
+from aiounifi.interfaces.devices import Devices
+from aiounifi.models.device import Device
+from aiounifi.models.event import Event, EventKey
+
+from homeassistant.core import callback
+from homeassistant.helpers import entity_registry as er
+from homeassistant.helpers.dispatcher import async_dispatcher_connect
+from homeassistant.helpers.entity import DeviceInfo, Entity, EntityDescription
+
+from .controller import UniFiController
+
+DataT = TypeVar("DataT", bound=Device)
+HandlerT = TypeVar("HandlerT", bound=Devices)
+SubscriptionT = Callable[[CallbackType, ItemEvent], UnsubscribeType]
+
+
+@dataclass
+class UnifiDescription(Generic[HandlerT, DataT]):
+    """Validate and load entities from different UniFi handlers."""
+
+    allowed_fn: Callable[[UniFiController, str], bool]
+    api_handler_fn: Callable[[aiounifi.Controller], HandlerT]
+    available_fn: Callable[[UniFiController, str], bool]
+    device_info_fn: Callable[[aiounifi.Controller, str], DeviceInfo | None]
+    event_is_on: tuple[EventKey, ...] | None
+    event_to_subscribe: tuple[EventKey, ...] | None
+    name_fn: Callable[[DataT], str | None]
+    object_fn: Callable[[aiounifi.Controller, str], DataT]
+    supported_fn: Callable[[UniFiController, str], bool | None]
+    unique_id_fn: Callable[[UniFiController, str], str]
+
+
+@dataclass
+class UnifiEntityDescription(EntityDescription, UnifiDescription[HandlerT, DataT]):
+    """UniFi Entity Description."""
+
+
+class UnifiEntity(Entity, Generic[HandlerT, DataT]):
```

Comment:
Should this be an ABC?

Reply:
Haven't used ABC can you elaborate?

Reply:
My bad, since `UnifiEntity` is derived from `Entity` it already inherits `ABC`.